include/spblas/vendor/armpl/multiply_impl.hpp

```diff
@@ -0,0 +1,147 @@
+               __mdspan::layout_right> &&
+           std::is_same_v<typename std::remove_cvref_t<C>::layout_type,
+                          __mdspan::layout_right>
+void multiply(A&& a, B&& b, C&& c) {
+  auto a_base = __detail::get_ultimate_base(a);
+  auto b_base = __detail::get_ultimate_base(b);
+
+  auto alpha_optional = __detail::get_scaling_factor(a, b);
+  tensor_scalar_t<A> alpha = alpha_optional.value_or(1);
+
+  armpl_spmat_t a_handle, b_handle, c_handle;
+
+  __armpl::create_spmat_csr<tensor_scalar_t<A>>(
+      &a_handle, __backend::shape(a_base)[0], __backend::shape(a_base)[1],
+      a_base.rowptr().data(), a_base.colind().data(), a_base.values().data(),
+      ARMPL_SPARSE_CREATE_NOCOPY);
+
+  __armpl::create_spmat_dense<tensor_scalar_t<B>>(
+      &b_handle, ARMPL_ROW_MAJOR, __backend::shape(b_base)[0],
+      __backend::shape(b_base)[1], __backend::shape(b_base)[1],
+      b_base.data_handle(), ARMPL_SPARSE_CREATE_NOCOPY);
+
+  __armpl::create_spmat_dense<tensor_scalar_t<C>>(
+      &c_handle, ARMPL_ROW_MAJOR, __backend::shape(c)[0],
+      __backend::shape(c)[1], __backend::shape(c)[1], c.data_handle(),
+      ARMPL_SPARSE_CREATE_NOCOPY);
+
+  __armpl::spmm_exec<tensor_scalar_t<A>>(ARMPL_SPARSE_OPERATION_NOTRANS,
+                                         ARMPL_SPARSE_OPERATION_NOTRANS, alpha,
+                                         a_handle, b_handle, 0, c_handle);
+
+  armpl_int_t m, n;
+  tensor_scalar_t<C>* armpl_values;
+  __armpl::export_spmat_dense<tensor_scalar_t<C>>(c_handle, ARMPL_ROW_MAJOR, &m,
+                                                  &n, &armpl_values);
+
+  std::copy(armpl_values, armpl_values + (m * n), c.data_handle());
+
+  free(armpl_values);
+
+  armpl_spmat_destroy(a_handle);
+  armpl_spmat_destroy(b_handle);
+  armpl_spmat_destroy(c_handle);
+}
+
+template <matrix A, matrix B, matrix C>
+  requires __detail::has_csr_base<A> && __detail::has_csr_base<B> &&
+           __detail::is_csr_view_v<C>
+operation_info_t multiply_inspect(A&& a, B&& b, C&& c) {
+  auto a_base = __detail::get_ultimate_base(a);
+  auto b_base = __detail::get_ultimate_base(b);
+
+  auto alpha_optional = __detail::get_scaling_factor(a, b);
+  tensor_scalar_t<A> alpha = alpha_optional.value_or(1);
+
+  armpl_spmat_t a_handle, b_handle, c_handle;
+
+  __armpl::create_spmat_csr<tensor_scalar_t<A>>(
+      &a_handle, __backend::shape(a_base)[0], __backend::shape(a_base)[1],
+      a_base.rowptr().data(), a_base.colind().data(), a_base.values().data(),
+      ARMPL_SPARSE_CREATE_NOCOPY);
+
+  __armpl::create_spmat_csr<tensor_scalar_t<B>>(
+      &b_handle, __backend::shape(b_base)[0], __backend::shape(b_base)[1],
+      b_base.rowptr().data(), b_base.colind().data(), a_base.values().data(),
+      ARMPL_SPARSE_CREATE_NOCOPY);
+
+  c_handle =
+      armpl_spmat_create_null(__backend::shape(c)[0], __backend::shape(c)[1]);
+
+  __armpl::spmm_exec<tensor_scalar_t<A>>(ARMPL_SPARSE_OPERATION_NOTRANS,
```

Comment:
A better alternative to doing spmm during `inspect` in order to query `m`, `n` and `nnz` is to call [armpl_spmat_hint](https://developer.arm.com/documentation/101004/latest/Sparse-Linear-Algebra/Sparse-Linear-Algebra-Functions/armpl-spmat-hint?lang=en) on `c_handle` using the `ARMPL_SPARSE_SPMM_STRAT_OPT_PART_STRUCT` hint, to set a hint that we want to compute part of the structure of the output during `optimize`. And then call [armpl_spmm_optimize](https://developer.arm.com/documentation/101004/2404/Sparse-Linear-Algebra/Sparse-Linear-Algebra-Functions/armpl-spmm-optimize?lang=en),  before query. This will give you the information needed without doing a full multiply here.

Note that users only have to state whether their scalars are zero, one or not-zero-or-one at this stage as well: that's a hint too.

```
auto alpha_hint = alpha == (T)0 ? ARMPL_SPARSE_SCALAR_ZERO : alpha == (T)1 ? ARMPL_SPARSE_SCALAR_ONE : ARMPL_SPARSE_SCALAR_ANY;

armpl_spmat_hint(C, ARMPL_SPARSE_HINT_SPMM_STRATEGY, ARMPL_SPARSE_SPMM_STRAT_OPT_PART_STRUCT);

armpl_spmm_optimize(ARMPL_SPARSE_OPERATION_NOTRANS, ARMPL_SPARSE_OPERATION_NOTRANS, alpha_hint, a_handle, b_handle, beta_hint, c_handle));
```

Reply:
Do you want to take a swing at adding this to the ArmPL backend?  I believe you should be able to push commits directly to the `armpl` branch.

(In my understanding, this doesn't fix the issue with having to perform a copy when exporting from the ArmPL matrix, but it does make `inspect` behave more like a user might expect in that it doesn't do the whole operation.)

Reply:
👍 Absolutely, will do.

Yes, your understanding is correct - this is just to fix `inspect`. I think we can avoid those copies in `execute` in the longer term though.

Reply:
I had a play around with this and it turns out not to be so simple, basically because our optimize is not guaranteed to do anything since it's based on ignorable hints... in this case, if B's column indices are not sorted, then we don't get any structural info - even NNZ. I uploaded a patch to illustrate, but feel free to back this out.

Reply:
Ah, gotcha.  This is good to know---this is the primary motivation for further splitting the phases up into `inspect`, `execute`, and `extract` like I mentioned in my last email to the group.  You do ArmPL's optimize in `inspect`, but it's not required to return any information.  Only after you hit `execute` and have the computed result in some internal format do you return the information about the output, which you can then use to ensure you have enough space for `extract`.

I'll add a commit to revert your changes for now, but I'll leave your commits in the branch history as it will be useful to have in the future.